# Customer Value Radar — Audit qualité

Audit Qualité du dataset Online Retail II

**Source**
- Online Retail II — UCI Machine Learning Repository
- Période : 01/12/2009 au 09/12/2011
- Volume : **1 067 371 observations**
- E-commerçant britannique sans magasin physique
- De nombreux clients sont des grossistes
- lien Source du dataset : https://archive.ics.uci.edu/dataset/502/online+retail+ii

**Définitions UCI**
- `Invoice` : identifiant de facture ; préfixe `C` = annulation (cancellation)
- `StockCode` : code produit
- `Description` : nom du produit
- `Quantity` : quantité
- `InvoiceDate` : date et heure de la transaction
- `Price` : prix unitaire en livres sterling £
- `Customer ID` : ID client
- `Country` : pays de résidence du client





In [1]:
from pathlib import Path

import pandas as pd


In [2]:
# Dataset
DATA_PATH = Path("../data/raw/online_retail_II.xlsx")

print(DATA_PATH.exists())
print(DATA_PATH)

True
../data/raw/online_retail_II.xlsx


In [3]:
# Feuilles EXCEL

excel_file = pd.ExcelFile(DATA_PATH)

excel_file.sheet_names

['Year 2009-2010', 'Year 2010-2011']

In [4]:
# Structure des feuilles excel

for sheet in excel_file.sheet_names:
    print(f"\n--- {sheet} ---")
    
    sample = pd.read_excel(
        DATA_PATH,
        sheet_name=sheet,
        nrows=5
    )
    
    display(sample)


--- Year 2009-2010 ---


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085,United Kingdom



--- Year 2010-2011 ---


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom


In [5]:
# Les colonnes sont : 

for sheet in excel_file.sheet_names:
    sample = pd.read_excel(
        DATA_PATH,
        sheet_name=sheet,
        nrows=5
    )
    
    print(f"\n{sheet}")
    print(sample.columns.tolist())


Year 2009-2010
['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'Price', 'Customer ID', 'Country']

Year 2010-2011
['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'Price', 'Customer ID', 'Country']


In [6]:
excel_file.sheet_names

['Year 2009-2010', 'Year 2010-2011']

In [7]:
sample.columns.tolist()

['Invoice',
 'StockCode',
 'Description',
 'Quantity',
 'InvoiceDate',
 'Price',
 'Customer ID',
 'Country']

In [8]:
# Chargement du dataset

df_2009_2010 = pd.read_excel(
    DATA_PATH,
    sheet_name="Year 2009-2010"
)

df_2010_2011 = pd.read_excel(
    DATA_PATH,
    sheet_name="Year 2010-2011"
)

print("2009-2010 :", df_2009_2010.shape)
print("2010-2011 :", df_2010_2011.shape)

2009-2010 : (525461, 8)
2010-2011 : (541910, 8)


In [9]:
# Data type

print(" 2009-2010 ")
df_2009_2010.info()

print("\n 2010-2011 ")
df_2010_2011.info()

 2009-2010 
<class 'pandas.DataFrame'>
RangeIndex: 525461 entries, 0 to 525460
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      525461 non-null  object        
 1   StockCode    525461 non-null  object        
 2   Description  522533 non-null  object        
 3   Quantity     525461 non-null  int64         
 4   InvoiceDate  525461 non-null  datetime64[us]
 5   Price        525461 non-null  float64       
 6   Customer ID  417534 non-null  float64       
 7   Country      525461 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 32.1+ MB

 2010-2011 
<class 'pandas.DataFrame'>
RangeIndex: 541910 entries, 0 to 541909
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      541910 non-null  object        
 1   StockCode    541910 non-null  ob

In [10]:
# Valeurs manquantes :

print(" Valeurs manquantes 2009-2010 ")
display(df_2009_2010.isna().sum())

print(" Valeurs manquantes 2010-2011 ")
display(df_2010_2011.isna().sum())

 Valeurs manquantes 2009-2010 


Invoice             0
StockCode           0
Description      2928
Quantity            0
InvoiceDate         0
Price               0
Customer ID    107927
Country             0
dtype: int64

 Valeurs manquantes 2010-2011 


Invoice             0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
Price               0
Customer ID    135080
Country             0
dtype: int64

Beaucoup de Customer ID manquants : On les gardes pour analyser les transactions mais on ne peut pas les garder pour des analyses clients .

In [11]:
# Verification des periodes : 

print("2009-2010")
print("Date minimum :", df_2009_2010["InvoiceDate"].min())
print("Date maximum :", df_2009_2010["InvoiceDate"].max())

print("\n2010-2011")
print("Date minimum :", df_2010_2011["InvoiceDate"].min())
print("Date maximum :", df_2010_2011["InvoiceDate"].max())

2009-2010
Date minimum : 2009-12-01 07:45:00
Date maximum : 2010-12-09 20:01:00

2010-2011
Date minimum : 2010-12-01 08:26:00
Date maximum : 2011-12-09 12:50:00


Il y a un chevauchement des dates du 1er au 9 décembre 2010.
Avant de concaténer les deux feuilles il faut verifier si : 
- Les transactions apparaissent deux fois ?
- Transactions différentes ? 
- Particularité du fichier ?

In [12]:
# Controle des infos : 
for name, df in [
    ("2009-2010", df_2009_2010),
    ("2010-2011", df_2010_2011)
]:
    print(f"\n {name} ")

    print("Nombre de lignes :", len(df))
    print("Factures distinctes :", df["Invoice"].nunique())
    print("Produits distincts :", df["StockCode"].nunique())
    print("Clients distincts :", df["Customer ID"].nunique())
    print("Pays distincts :", df["Country"].nunique())

    print("Quantité minimum :", df["Quantity"].min())
    print("Quantité maximum :", df["Quantity"].max())

    print("Prix minimum :", df["Price"].min())
    print("Prix maximum :", df["Price"].max())


 2009-2010 
Nombre de lignes : 525461
Factures distinctes : 28816
Produits distincts : 4632
Clients distincts : 4383
Pays distincts : 40
Quantité minimum : -9600
Quantité maximum : 19152
Prix minimum : -53594.36
Prix maximum : 25111.09

 2010-2011 
Nombre de lignes : 541910
Factures distinctes : 25900
Produits distincts : 4070
Clients distincts : 4372
Pays distincts : 38
Quantité minimum : -80995
Quantité maximum : 80995
Prix minimum : -11062.06
Prix maximum : 38970.0


Un prix unitaire negatif n'est pas normale.
et quantité négatif non plus. 
On va voir plus en détail ce a quoi c'est du.

In [13]:
# On conserve la provenance de chaque ligne avant de réunir les deux feuilles.
# On ajoute SourcePeriod pour savoir de quelle feuille originale vient une ligne pour la retraçer.

df_2009_2010["SourcePeriod"] = "2009-2010"
df_2010_2011["SourcePeriod"] = "2010-2011"

df_raw = pd.concat(
    [df_2009_2010, df_2010_2011],
    ignore_index=True
)

print("Dimensions du dataset complet :", df_raw.shape)

Dimensions du dataset complet : (1067371, 9)


In [14]:
# Verification des doublons : 

duplicate_count = df_raw.duplicated().sum()

print("Nombre de doublons exacts :", duplicate_count)

Nombre de doublons exacts : 12133


In [15]:
duplicate_rows = df_raw[df_raw.duplicated(keep=False)]

display(duplicate_rows.head(20))

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,SourcePeriod
362,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom,2009-2010
363,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom,2009-2010
365,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom,2009-2010
367,489517,22319,HAIRCLIPS FORTIES FABRIC ASSORTED,12,2009-12-01 11:34:00,0.65,16329.0,United Kingdom,2009-2010
368,489517,22130,PARTY CONE CHRISTMAS DECORATION,6,2009-12-01 11:34:00,0.85,16329.0,United Kingdom,2009-2010
371,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom,2009-2010
379,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom,2009-2010
383,489517,22130,PARTY CONE CHRISTMAS DECORATION,6,2009-12-01 11:34:00,0.85,16329.0,United Kingdom,2009-2010
384,489517,22319,HAIRCLIPS FORTIES FABRIC ASSORTED,12,2009-12-01 11:34:00,0.65,16329.0,United Kingdom,2009-2010
385,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom,2009-2010


In [16]:
business_columns = [
    "Invoice",
    "StockCode",
    "Description",
    "Quantity",
    "InvoiceDate",
    "Price",
    "Customer ID",
    "Country"
]

duplicates_business = df_raw.duplicated(
    subset=business_columns,
    keep=False
)

print(
    "Lignes potentiellement dupliquées sur les colonnes métier :",
    duplicates_business.sum()
)

Lignes potentiellement dupliquées sur les colonnes métier : 67242


In [17]:
display(
    df_raw.loc[duplicates_business]
    .sort_values(["InvoiceDate", "Invoice", "StockCode"])
    .head(30)
)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,SourcePeriod
379,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom,2009-2010
391,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom,2009-2010
365,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom,2009-2010
386,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom,2009-2010
363,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom,2009-2010
371,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom,2009-2010
394,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom,2009-2010
362,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom,2009-2010
385,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom,2009-2010
368,489517,22130,PARTY CONE CHRISTMAS DECORATION,6,2009-12-01 11:34:00,0.85,16329.0,United Kingdom,2009-2010


In [18]:
# Verification chevauchement dates décembre 2010

overlap_2009_2010 = df_2009_2010[
    df_2009_2010["InvoiceDate"].between(
        "2010-12-01",
        "2010-12-09 23:59:59"
    )
]

overlap_2010_2011 = df_2010_2011[
    df_2010_2011["InvoiceDate"].between(
        "2010-12-01",
        "2010-12-09 23:59:59"
    )
]

print("Lignes décembre 2010 dans feuille 2009-2010 :", len(overlap_2009_2010))
print("Lignes décembre 2010 dans feuille 2010-2011 :", len(overlap_2010_2011))

Lignes décembre 2010 dans feuille 2009-2010 : 22523
Lignes décembre 2010 dans feuille 2010-2011 : 22523


In [19]:
print("Factures feuille 2009-2010 :", overlap_2009_2010["Invoice"].nunique())

print("Factures feuille 2010-2011 :",overlap_2010_2011["Invoice"].nunique())

Factures feuille 2009-2010 : 1088
Factures feuille 2010-2011 : 1088


In [20]:
common_invoices = set(overlap_2009_2010["Invoice"]).intersection(
    set(overlap_2010_2011["Invoice"])
)

print("Factures présentes dans les deux feuilles :", len(common_invoices))

Factures présentes dans les deux feuilles : 1088


Sur la periode les factures apparaissent sur les deux feuilles.

In [21]:
# Colonnes à comparer entre les deux feuilles

compare_columns = [
    "Invoice",
    "StockCode",
    "Description",
    "Quantity",
    "InvoiceDate",
    "Price",
    "Customer ID",
    "Country"
]

# On trie les deux périodes 
overlap_a = (
    overlap_2009_2010[compare_columns]
    .sort_values(compare_columns, na_position="last")
    .reset_index(drop=True)
)

overlap_b = (
    overlap_2010_2011[compare_columns]
    .sort_values(compare_columns, na_position="last")
    .reset_index(drop=True)
)

print("Même nombre de lignes :", len(overlap_a) == len(overlap_b))
print("Contenu strictement identique :", overlap_a.equals(overlap_b))

Même nombre de lignes : True
Contenu strictement identique : True


### Constat de qualité : 
- les deux feuilles couvrent une periode commune du 1er au 9 décembre 2010
- Cette période contient 22 523 lignes et 1 088 factures dans chacune des deux feuilles. 
- Les 1 088 numéros de facture sont présents dans les deux sources.

Vérification pour voir si les lignes sont strictement identiques

In [22]:
overlap_a.equals(overlap_b)

True

### Conclusion 

La période du 1er au 9 décembre 2010 est présente dans les deux feuilles du fichier original.

Les deux périodes contiennent exactement :

- 22 523 lignes ;
- 1 088 factures distinctes.

Après tri sur l’ensemble des colonnes métier, les deux sous-ensembles sont strictement identiques.


lors de l’intégration des deux feuilles, une seule occurrence de cette période sera conservée afin d’éviter un double comptage des transactions.

La feuille Year 2010-2011 sera conservée comme source de référence pour cette période de chevauchement.

### Comprendre les quantités negatives : 

In [23]:
# Analyse des quantités négatives

negative_quantity = df_raw[df_raw["Quantity"] < 0]

print("Nombre de lignes avec quantité négative :", len(negative_quantity))
print("Nombre de factures concernées :", negative_quantity["Invoice"].nunique())

Nombre de lignes avec quantité négative : 22950
Nombre de factures concernées : 11684


In [24]:
display(
    negative_quantity[
        [
            "Invoice",
            "StockCode",
            "Description",
            "Quantity",
            "Price",
            "Customer ID",
            "Country",
            "InvoiceDate"
        ]
    ].head(30)
)

,Invoice,StockCode,Description,Quantity,Price,Customer ID,Country,InvoiceDate
178,C489449,22087,PAPER BUNTING WHITE LACE,-12,2.95,16321.0,Australia,2009-12-01 10:33:00
179,C489449,85206A,CREAM FELT EASTER EGG BASKET,-6,1.65,16321.0,Australia,2009-12-01 10:33:00
180,C489449,21895,POTTING SHED SOW 'N' GROW SET,-4,4.25,16321.0,Australia,2009-12-01 10:33:00
181,C489449,21896,POTTING SHED TWINE,-6,2.10,16321.0,Australia,2009-12-01 10:33:00
182,C489449,22083,PAPER CHAIN KIT RETRO SPOT,-12,2.95,16321.0,Australia,2009-12-01 10:33:00
183,C489449,21871,SAVE THE PLANET MUG,-12,1.25,16321.0,Australia,2009-12-01 10:33:00
184,C489449,84946,ANTIQUE SILVER TEA GLASS ETCHED,-12,1.25,16321.0,Australia,2009-12-01 10:33:00
185,C489449,84970S,HANGING HEART ZINC T-LIGHT HOLDER,-24,0.85,16321.0,Australia,2009-12-01 10:33:00
186,C489449,22090,PAPER BUNTING RETRO SPOTS,-12,2.95,16321.0,Australia,2009-12-01 10:33:00
196,C489459,90200A,PURPLE SWEETHEART BRACELET,-3,4.25,17592.0,United Kingdom,2009-12-01 10:44:00


In [25]:
# Verification des factures : 

negative_invoice_check = (
    negative_quantity["Invoice"]
    .astype(str)
    .str.startswith("C")
    .value_counts()
)

display(negative_invoice_check)

Invoice
True     19493
False     3457
Name: count, dtype: int64

In [26]:
# Calculons les proportions : 

total_negative = len(negative_quantity)

negative_with_c = (
    negative_quantity["Invoice"]
    .astype(str)
    .str.startswith("C")
    .sum()
)

negative_without_c = total_negative - negative_with_c

print("Quantités négatives :", total_negative)
print("Avec facture C :", negative_with_c)
print("Sans facture C :", negative_without_c)

print(
    "Part avec facture C :",
    round(negative_with_c / total_negative * 100, 2),
    "%"
)

Quantités négatives : 22950
Avec facture C : 19493
Sans facture C : 3457
Part avec facture C : 84.94 %


## Construction du dataset consolidé

On a chevauchement de 22 523 lignes entre les deux feuilles pour la période du 1er au 9 décembre 2010.

Afin d'éviter tout doublons, une seule occurrence de cette période est conservée.

Le dataset consolidé contient :
- la feuille 2009-2010 jusqu'au 30 novembre 2010 ;
- l'intégralité de la feuille 2010-2011.


In [27]:
# Retrait de la période dupliquée dans la première feuille

df_2009_2010_unique = df_2009_2010[
    df_2009_2010["InvoiceDate"] < "2010-12-01"
].copy()

# Consolidation des deux périodes

df_consolidated = pd.concat(
    [
        df_2009_2010_unique,
        df_2010_2011
    ],
    ignore_index=True
)

print("Dataset brut :", len(df_raw))
print("Dataset consolidé :", len(df_consolidated))
print("Lignes retirées :", len(df_raw) - len(df_consolidated))

Dataset brut : 1067371
Dataset consolidé : 1044848
Lignes retirées : 22523


In [28]:
# verification de la periode : 
print("Date minimum :", df_consolidated["InvoiceDate"].min())
print("Date maximum :", df_consolidated["InvoiceDate"].max())

Date minimum : 2009-12-01 07:45:00
Date maximum : 2011-12-09 12:50:00


In [29]:
# Recalcul des quantités negative :
negative_quantity = df_consolidated[
    df_consolidated["Quantity"] < 0
]

print(
    "Nombre de lignes avec quantité négative :",
    len(negative_quantity)
)

print(
    "Nombre de factures concernées :",
    negative_quantity["Invoice"].nunique()
)

Nombre de lignes avec quantité négative : 22557
Nombre de factures concernées : 11684


In [30]:
total_negative = len(negative_quantity)

negative_with_c = (
    negative_quantity["Invoice"]
    .astype(str)
    .str.startswith("C")
    .sum()
)

negative_without_c = total_negative - negative_with_c

print("Quantités négatives :", total_negative)
print("Avec facture C :", negative_with_c)
print("Sans facture C :", negative_without_c)

print(
    "Part avec facture C :",
    round(negative_with_c / total_negative * 100, 2),
    "%"
)

Quantités négatives : 22557
Avec facture C : 19164
Sans facture C : 3393
Part avec facture C : 84.96 %


### constat sur les quantités négatives

Après suppression du chevauchement entre les deux feuilles:
le dataset consolidé contient 22 557 lignes avec une quantité négative, 
réparties sur 11 684 factures.

Parmi ces lignes :

- 19 164, soit 84,96 %, appartiennent à une facture dont l'identifiant commence par `C` ;
- 3 393, soit 15,04 %, possèdent une quantité négative sans facture commençant par `C`.

Les quantités négatives ne peuvent donc pas être assimilées automatiquement à des annulations clients.

Les 3 393 cas restants doivent être analysés avant de définir une règle de nettoyage.

In [31]:
# On isole les cas sans le C : 
negative_without_c_df = negative_quantity[
    ~negative_quantity["Invoice"]
    .astype(str)
    .str.startswith("C")
].copy()

print(
    "Nombre de lignes avec quantité négative sans facture C :",
    len(negative_without_c_df)
)

Nombre de lignes avec quantité négative sans facture C : 3393


In [32]:
# Descriptions : 
display(
    negative_without_c_df["Description"]
    .value_counts(dropna=False)
    .head(30)
)

Description
NaN                       2633
check                      121
damages                     83
?                           81
damaged                     78
missing                     27
sold as set on dotcom       20
Damaged                     17
smashed                      9
thrown away                  9
Unsaleable, destroyed.       9
dotcom                       8
damages?                     7
??                           7
crushed                      6
given away                   6
MIA                          5
Damages                      5
counted                      5
checked                      5
wet damaged                  5
ebay                         5
ebay sales                   4
No Stock                     3
Dotcom                       3
broken                       3
CHECK                        3
wet pallet                   3
lost                         2
wet                          2
Name: count, dtype: int64

In [33]:
display(
    negative_without_c_df["StockCode"]
    .astype(str)
    .value_counts()
    .head(30)
)

StockCode
22423      11
82494L      6
22719       6
46000M      5
84016       5
85017A      5
47566B      5
20852       5
85175       5
21830       5
71477       4
46000S      4
21768       4
84559D      4
84990       4
35972       4
79000       4
20713       4
21843       4
84406B      4
85172       4
72802C      4
79323GR     3
21098       3
85064       3
21493       3
21489       3
21490       3
51008       3
79030A      3
Name: count, dtype: int64

In [34]:
display(
    negative_without_c_df[
        [
            "Invoice",
            "StockCode",
            "Description",
            "Quantity",
            "Price",
            "Customer ID",
            "Country",
            "InvoiceDate"
        ]
    ].head(50)
)

,Invoice,StockCode,Description,Quantity,Price,Customer ID,Country,InvoiceDate
263,489464,21733,85123a mixed,-96,0.0,NaN,United Kingdom,2009-12-01 10:52:00
283,489463,71477,short,-240,0.0,NaN,United Kingdom,2009-12-01 10:52:00
284,489467,85123A,21733 mixed,-192,0.0,NaN,United Kingdom,2009-12-01 10:53:00
470,489521,21646,NaN,-50,0.0,NaN,United Kingdom,2009-12-01 11:44:00
3114,489655,20683,NaN,-44,0.0,NaN,United Kingdom,2009-12-01 17:26:00
3162,489660,35956,lost,-1043,0.0,NaN,United Kingdom,2009-12-01 17:43:00
3168,489663,35605A,damages,-117,0.0,NaN,United Kingdom,2009-12-01 18:02:00
4296,489806,18010,NaN,-770,0.0,NaN,United Kingdom,2009-12-02 12:42:00
4538,489820,21133,invcd as 84879?,-720,0.0,NaN,United Kingdom,2009-12-02 13:23:00
4566,489821,85049G,NaN,-240,0.0,NaN,United Kingdom,2009-12-02 13:25:00


In [35]:
display(
    negative_without_c_df[
        [
            "Invoice",
            "StockCode",
            "Description",
            "Quantity",
            "Price",
            "Customer ID",
            "Country",
            "InvoiceDate"
        ]
    ]
    .sort_values("Quantity")
    .head(30)
)

,Invoice,StockCode,Description,Quantity,Price,Customer ID,Country,InvoiceDate
728468,556691,23005,printing smudges/thrown away,-9600,0.0,NaN,United Kingdom,2011-06-14 10:37:00
728467,556690,23005,printing smudges/thrown away,-9600,0.0,NaN,United Kingdom,2011-06-14 10:37:00
303996,519017,22759,NaN,-9600,0.0,NaN,United Kingdom,2010-08-13 09:14:00
156488,504311,22197,NaN,-9200,0.0,NaN,United Kingdom,2010-04-12 14:39:00
728466,556687,23003,Printing smudges/thrown away,-9058,0.0,NaN,United Kingdom,2011-06-14 10:36:00
428975,530348,16235,?,-9000,0.0,NaN,United Kingdom,2010-11-02 15:48:00
194372,507913,10120,Zebra invcing error,-9000,0.0,NaN,United Kingdom,2010-05-11 17:16:00
437008,531177,20852,Given away,-8800,0.0,NaN,United Kingdom,2010-11-05 14:40:00
258814,514363,84016,Ebay sales by the box.,-7100,0.0,NaN,United Kingdom,2010-07-01 18:07:00
269558,515462,20852,given away,-7000,0.0,NaN,United Kingdom,2010-07-12 15:47:00


### on peut distinguer : 

Quantity > 0
vente / mouvement positif

Quantity < 0 + Invoice commence par C
annulation / retour commercial identifié

Quantity < 0 + Invoice ne commence pas par C
mouvement négatif non commercial ou non clairement identifié
(casse, perte, correction, mise au rebut, erreur d'inventaire...)

### Conclusion sur les quantités négatives

L'analyse met en évidence deux types principaux de mouvements négatifs.

Sur les 22 557 lignes présentant une quantité négative :

- 19 164 lignes (84,96 %) appartiennent à une facture commençant par `C` ;
- 3 393 lignes (15,04 %) ne sont pas associées à une facture commençant par `C`.

Les factures commençant par `C` seront considérées comme des annulations ou retours commerciaux.

Les mouvements négatifs sans facture `C` présentent notamment des descriptions telles que `damages`, `damaged`, `missing`, `lost`, `smashed`, `thrown away`, `Given away` ou encore des mentions d'erreurs de facturation.

Une part importante de ces lignes ne possède aucune description.

### Décision :

- les mouvements négatifs seront conservés dans les données transformées afin de préserver l'information source 
- les factures `C` seront identifiées séparément comme annulations/retours commerciaux 
- les quantités négatives hors facture `C` seront identifiées comme mouvements négatifs non commerciaux ou non déterminés 
- ces mouvements ne seront pas intégrés aux indicateurs de ventes sans traitement spécifique.

### Analysons maintenant Les Prix (Price)

In [36]:
# Prix négatifs : 
negative_price = df_consolidated[
    df_consolidated["Price"] < 0
].copy()

print("Nombre de lignes avec prix négatif :", len(negative_price))

Nombre de lignes avec prix négatif : 5


In [37]:
display(
    negative_price[
        [
            "Invoice",
            "StockCode",
            "Description",
            "Quantity",
            "Price",
            "Customer ID",
            "Country",
            "InvoiceDate"
        ]
    ].sort_values("Price")
)

,Invoice,StockCode,Description,Quantity,Price,Customer ID,Country,InvoiceDate
179403,A506401,B,Adjust bad debt,1,-53594.36,NaN,United Kingdom,2010-04-29 13:36:00
276274,A516228,B,Adjust bad debt,1,-44031.79,NaN,United Kingdom,2010-07-19 11:24:00
403472,A528059,B,Adjust bad debt,1,-38925.87,NaN,United Kingdom,2010-10-20 12:04:00
802921,A563186,B,Adjust bad debt,1,-11062.06,NaN,United Kingdom,2011-08-12 14:51:00
802922,A563187,B,Adjust bad debt,1,-11062.06,NaN,United Kingdom,2011-08-12 14:52:00


In [38]:
# Prix égale a 0

zero_price = df_consolidated[
    df_consolidated["Price"] == 0
].copy()

print("Nombre de lignes avec prix égal à 0 :", len(zero_price))

Nombre de lignes avec prix égal à 0 : 6024


In [39]:
display(
    zero_price[
        [
            "Invoice",
            "StockCode",
            "Description",
            "Quantity",
            "Price",
            "Customer ID",
            "Country",
            "InvoiceDate"
        ]
    ].head(50)
)

,Invoice,StockCode,Description,Quantity,Price,Customer ID,Country,InvoiceDate
263,489464,21733,85123a mixed,-96,0.0,NaN,United Kingdom,2009-12-01 10:52:00
283,489463,71477,short,-240,0.0,NaN,United Kingdom,2009-12-01 10:52:00
284,489467,85123A,21733 mixed,-192,0.0,NaN,United Kingdom,2009-12-01 10:53:00
470,489521,21646,NaN,-50,0.0,NaN,United Kingdom,2009-12-01 11:44:00
3114,489655,20683,NaN,-44,0.0,NaN,United Kingdom,2009-12-01 17:26:00
3161,489659,21350,NaN,230,0.0,NaN,United Kingdom,2009-12-01 17:39:00
3162,489660,35956,lost,-1043,0.0,NaN,United Kingdom,2009-12-01 17:43:00
3168,489663,35605A,damages,-117,0.0,NaN,United Kingdom,2009-12-01 18:02:00
3731,489781,84292,NaN,17,0.0,NaN,United Kingdom,2009-12-02 11:45:00
4296,489806,18010,NaN,-770,0.0,NaN,United Kingdom,2009-12-02 12:42:00


In [40]:
display(
    zero_price["Description"]
    .value_counts(dropna=False)
    .head(30)
)

Description
NaN                                    4275
check                                   160
?                                        90
damages                                  83
damaged                                  81
found                                    28
missing                                  27
sold as set on dotcom                    20
Damaged                                  17
adjustment                               16
OWL DOORSTOP                             14
POLYESTER FILLER PAD 45x45cm             12
dotcom                                   12
POLYESTER FILLER PAD 40x40cm             10
smashed                                   9
Found                                     9
FRENCH BLUE METAL DOOR SIGN 1             9
thrown away                               9
Unsaleable, destroyed.                    9
PICNIC BASKET WICKER LARGE                8
AIRLINE BAG VINTAGE WORLD CHAMPION        8
checked                                   8
BOX OF 24 COCKTAIL P

In [41]:
# Prix très élevées : 
positive_price = df_consolidated[
    df_consolidated["Price"] > 0
]["Price"]

print(positive_price.describe(
    percentiles=[0.50, 0.90, 0.95, 0.99, 0.999]
))

count    1.038819e+06
mean     4.769935e+00
std      9.257830e+01
min      1.000000e-03
50%      2.100000e+00
90%      7.950000e+00
95%      9.950000e+00
99%      1.800000e+01
99.9%    2.165137e+02
max      3.897000e+04
Name: Price, dtype: float64


### Prix négatifs

Le dataset consolidé contient seulement 5 lignes avec un prix négatif.

Ces 5 lignes possèdent le `StockCode` B et la description `Adjust bad debt`.
Elles correspondent à des ajustements comptables et non à des ventes de produits.

**Décision :**
ces lignes seront conservées dans la couche de données nettoyée pour assurer la traçabilité, mais identifiées comme mouvements non commerciaux et exclues des indicateurs de ventes.

In [42]:
high_price_rows = (
    df_consolidated[
        df_consolidated["Price"] > 0
    ][
        [
            "Invoice",
            "StockCode",
            "Description",
            "Quantity",
            "Price",
            "Customer ID",
            "Country",
            "InvoiceDate"
        ]
    ]
    .sort_values("Price", ascending=False)
    .head(30)
)

display(high_price_rows)

,Invoice,StockCode,Description,Quantity,Price,Customer ID,Country,InvoiceDate
725619,C556445,M,Manual,-1,38970.00,15098.0,United Kingdom,2011-06-10 15:31:00
241827,512771,M,Manual,1,25111.09,NaN,United Kingdom,2010-06-17 16:53:00
241824,C512770,M,Manual,-1,25111.09,17399.0,United Kingdom,2010-06-17 16:52:00
320581,C520667,BANK CHARGES,Bank Charges,-1,18910.69,NaN,United Kingdom,2010-08-27 13:42:00
1027540,C580605,AMAZONFEE,AMAZON FEE,-1,17836.46,NaN,United Kingdom,2011-12-05 11:36:00
546640,C540117,AMAZONFEE,AMAZON FEE,-1,16888.02,NaN,United Kingdom,2011-01-05 09:55:00
546641,C540118,AMAZONFEE,AMAZON FEE,-1,16453.71,NaN,United Kingdom,2011-01-05 09:57:00
519294,C537651,AMAZONFEE,AMAZON FEE,-1,13541.33,NaN,United Kingdom,2010-12-07 15:49:00
517954,C537630,AMAZONFEE,AMAZON FEE,-1,13541.33,NaN,United Kingdom,2010-12-07 15:04:00
517955,537632,AMAZONFEE,AMAZON FEE,1,13541.33,NaN,United Kingdom,2010-12-07 15:08:00


In [43]:
display(
    df_consolidated[
        df_consolidated["Price"] > 200
    ]["Description"]
    .value_counts(dropna=False)
    .head(30)
)

Description
Manual                                 468
DOTCOM POSTAGE                         467
AMAZON FEE                              35
Bank Charges                            27
VINTAGE RED KITCHEN CABINET             25
POSTAGE                                 24
VINTAGE BLUE KITCHEN CABINET            24
Discount                                12
CRUK Commission                         12
SAMPLES                                  7
FLAG OF ST GEORGE CAR FLAG               5
Adjustment by john on 26/01/2010 16      4
Adjustment by john on 26/01/2010 17      4
 Bank Charges                            3
Adjustment by Peter on 24/05/2010 1      2
Adjustment by Peter on Jun 25 2010       2
PICNIC BASKET WICKER 60 PIECES           2
Adjust bad debt                          1
Name: count, dtype: int64

In [44]:
print(
    "Nombre de lignes avec Price > 200 :",
    (df_consolidated["Price"] > 200).sum()
)

print(
    "Nombre de lignes avec Price > 1000 :",
    (df_consolidated["Price"] > 1000).sum()
)

print(
    "Nombre de lignes avec Price > 10000 :",
    (df_consolidated["Price"] > 10000).sum()
)

Nombre de lignes avec Price > 200 : 1124
Nombre de lignes avec Price > 1000 : 239
Nombre de lignes avec Price > 10000 : 23


### Conclusion sur les prix

L'analyse de la variable `Price` montre une distribution fortement asymétrique.

La majorité des prix observés est faible :

- médiane : £2,10 ;
- 90 % des lignes : prix inférieur ou égal à £7,95 
- 95 % : inférieur ou égal à £9,95 
- 99 % : inférieur ou égal à £18 
- 99,9 % : inférieur ou égal à environ £216,51

certaines valeurs sont très élevées :

- 1 124 lignes ont un prix supérieur à £200 
- 239 lignes ont un prix supérieur à £1 000 
- 23 lignes ont un prix supérieur à £10 000

Plusieurs de ces valeurs correspondent à des lignes (`Manual`, `DOTCOM POSTAGE`, `AMAZON FEE`, `Bank Charges`, ...), mais certains produits réels présentent également des prix élevés.

**Décision  :**
On ne supprime pas les valeurs extrèmes. 
Les lignes seront distinguées selon leur nature commerciale ou non commerciale.

In [45]:
high_price_1000 = df_consolidated[
    df_consolidated["Price"] > 1000
].copy()

display(
    high_price_1000["Description"]
    .value_counts(dropna=False)
    .head(30)
)

Description
Manual                                 166
AMAZON FEE                              33
DOTCOM POSTAGE                          29
Bank Charges                             2
Discount                                 2
POSTAGE                                  2
CRUK Commission                          2
Adjustment by john on 26/01/2010 17      1
FLAG OF ST GEORGE CAR FLAG               1
Adjust bad debt                          1
Name: count, dtype: int64

In [46]:
display(
    high_price_1000["StockCode"]
    .astype(str)
    .value_counts()
    .head(30)
)

StockCode
M               166
AMAZONFEE        33
DOT              29
BANK CHARGES      2
D                 2
POST              2
CRUK              2
ADJUST            1
84016             1
B                 1
Name: count, dtype: int64

In [47]:
display(
    high_price_1000[
        [
            "Invoice",
            "StockCode",
            "Description",
            "Quantity",
            "Price",
            "Customer ID",
            "Country",
            "InvoiceDate"
        ]
    ]
    .sort_values("Price", ascending=False)
    .head(30)
)

,Invoice,StockCode,Description,Quantity,Price,Customer ID,Country,InvoiceDate
725619,C556445,M,Manual,-1,38970.00,15098.0,United Kingdom,2011-06-10 15:31:00
241827,512771,M,Manual,1,25111.09,NaN,United Kingdom,2010-06-17 16:53:00
241824,C512770,M,Manual,-1,25111.09,17399.0,United Kingdom,2010-06-17 16:52:00
320581,C520667,BANK CHARGES,Bank Charges,-1,18910.69,NaN,United Kingdom,2010-08-27 13:42:00
1027540,C580605,AMAZONFEE,AMAZON FEE,-1,17836.46,NaN,United Kingdom,2011-12-05 11:36:00
546640,C540117,AMAZONFEE,AMAZON FEE,-1,16888.02,NaN,United Kingdom,2011-01-05 09:55:00
546641,C540118,AMAZONFEE,AMAZON FEE,-1,16453.71,NaN,United Kingdom,2011-01-05 09:57:00
517955,537632,AMAZONFEE,AMAZON FEE,1,13541.33,NaN,United Kingdom,2010-12-07 15:08:00
517954,C537630,AMAZONFEE,AMAZON FEE,-1,13541.33,NaN,United Kingdom,2010-12-07 15:04:00
519294,C537651,AMAZONFEE,AMAZON FEE,-1,13541.33,NaN,United Kingdom,2010-12-07 15:49:00


In [48]:
flag_high_price = high_price_1000[
    high_price_1000["Description"] == "FLAG OF ST GEORGE CAR FLAG"
]

display(
    flag_high_price[
        [
            "Invoice",
            "StockCode",
            "Description",
            "Quantity",
            "Price",
            "Customer ID",
            "Country",
            "InvoiceDate"
        ]
    ]
)

,Invoice,StockCode,Description,Quantity,Price,Customer ID,Country,InvoiceDate
192196,507637,84016,FLAG OF ST GEORGE CAR FLAG,1,1157.15,NaN,United Kingdom,2010-05-10 14:55:00


In [49]:
display(
    df_consolidated[
        df_consolidated["Description"] == "FLAG OF ST GEORGE CAR FLAG"
    ]["Price"].describe()
)

count      73.00000
mean       45.55411
std       177.99440
min         0.00000
25%         0.42000
50%         0.42000
75%         0.42000
max      1157.15000
Name: Price, dtype: float64

In [50]:
display(
    df_consolidated[
        df_consolidated["Description"] == "FLAG OF ST GEORGE CAR FLAG"
    ]["Price"]
    .value_counts()
    .sort_index()
)

Price
0.00        6
0.35        1
0.42       51
0.81        1
0.83        2
25.52       2
25.53       3
93.61       2
272.27      1
280.75      1
408.40      1
867.79      1
1157.15     1
Name: count, dtype: int64

### Conclusion sur les prix extrêmes

Une analyse spécifique a été réalisée sur la seule ligne supérieure à £1 000 dont la description correspondait à un produit : `FLAG OF ST GEORGE CAR FLAG`.

Le produit apparaît 73 fois dans le dataset.

Son prix médian est de £0,42 et 75 % des observations sont également à £0,42. Cependant, quelques valeurs très élevées sont observées, jusqu'à £1 157,15.

La valeur maximale est donc extrêmement éloignée du prix habituellement observé pour le même produit.

Cette analyse confirme qu'une valeur extrême ne doit pas être supprimée uniquement à partir d'un seuil global. La cohérence du prix doit également être étudiée par rapport à l'historique du produit.

**Décision :**
les valeurs extrêmes seront conservées dans les données transformées avec un indicateur de qualité. Elles pourront être exclues des KPI commerciaux lorsque leur caractère non commercial ou manifestement atypique est établi.

### Analyse des prix très élevés

L'analyse des 239 lignes dont le prix unitaire dépasse £1 000 montre que 238 correspondent à des lignes administratives, financières ou techniques.

Les principales catégories identifiées sont :

- `Manual` ;
- `AMAZON FEE` ;
- `DOTCOM POSTAGE` ;
- `Bank Charges` ;
- `Discount` ;
- `POSTAGE` ;
- `CRUK Commission` ;
- différents ajustements comptables.

Ces valeurs élevées ne représentent donc généralement pas le prix de vente d'un produit.

Une seule observation supérieure à £1 000 possède une description correspondant à un produit (`FLAG OF ST GEORGE CAR FLAG`). Cette observation fait l'objet d'un contrôle spécifique par comparaison avec les autres prix observés pour le même article.

**Décision :**
aucun seuil statistique arbitraire ne sera utilisé pour supprimer les prix élevés. La nature commerciale ou non commerciale de la ligne sera privilégiée.

## Analyse des identifiants clients manquants

In [51]:
missing_customer = df_consolidated[
    df_consolidated["Customer ID"].isna()
].copy()

total_rows = len(df_consolidated)
missing_customer_rows = len(missing_customer)

print("Nombre total de lignes :", total_rows)
print("Lignes sans Customer ID :", missing_customer_rows)

print(
    "Part du dataset sans Customer ID :",
    round(missing_customer_rows / total_rows * 100, 2),
    "%"
)

Nombre total de lignes : 1044848
Lignes sans Customer ID : 235287
Part du dataset sans Customer ID : 22.52 %


In [52]:
print(
    "Factures distinctes sans Customer ID :",
    missing_customer["Invoice"].nunique()
)

print(
    "Pays concernés :",
    missing_customer["Country"].nunique()
)

Factures distinctes sans Customer ID : 8752
Pays concernés : 15


In [53]:
display(
    missing_customer[
        [
            "Invoice",
            "StockCode",
            "Description",
            "Quantity",
            "Price",
            "Country",
            "InvoiceDate"
        ]
    ].head(30)
)

,Invoice,StockCode,Description,Quantity,Price,Country,InvoiceDate
263,489464,21733,85123a mixed,-96,0.00,United Kingdom,2009-12-01 10:52:00
283,489463,71477,short,-240,0.00,United Kingdom,2009-12-01 10:52:00
284,489467,85123A,21733 mixed,-192,0.00,United Kingdom,2009-12-01 10:53:00
470,489521,21646,NaN,-50,0.00,United Kingdom,2009-12-01 11:44:00
577,489525,85226C,BLUE PULL BACK RACING CAR,1,0.55,United Kingdom,2009-12-01 11:49:00
578,489525,85227,SET/6 3D KIT CARDS FOR KIDS,1,0.85,United Kingdom,2009-12-01 11:49:00
1055,489548,22271,FELTCRAFT DOLL ROSIE,1,2.95,United Kingdom,2009-12-01 12:32:00
1056,489548,22254,FELT TOADSTOOL LARGE,12,1.25,United Kingdom,2009-12-01 12:32:00
1057,489548,22273,FELTCRAFT DOLL MOLLY,3,2.95,United Kingdom,2009-12-01 12:32:00
1058,489548,22195,LARGE HEART MEASURING SPOONS,1,1.65,United Kingdom,2009-12-01 12:32:00


In [54]:
print(
    "Lignes sans client avec Quantity > 0 et Price > 0 :",
    (
        (missing_customer["Quantity"] > 0)
        & (missing_customer["Price"] > 0)
    ).sum()
)

Lignes sans client avec Quantity > 0 et Price > 0 : 228610


In [55]:
# Factures contenant au moins une ligne sans Customer ID

invoices_without_customer = set(
    missing_customer["Invoice"].unique()
)

# Factures contenant au moins une ligne avec Customer ID

invoices_with_customer = set(
    df_consolidated.loc[
        df_consolidated["Customer ID"].notna(),
        "Invoice"
    ].unique()
)

# Factures présentes dans les deux groupes

mixed_customer_invoices = (
    invoices_without_customer
    .intersection(invoices_with_customer)
)

print(
    "Factures avec lignes identifiées ET non identifiées :",
    len(mixed_customer_invoices)
)

Factures avec lignes identifiées ET non identifiées : 0


### Cohérence des Client ID au niveau des factures

Aucune facture ne contient simultanément des lignes avec `Customer ID` renseigné et des lignes sans `Customer ID`.

L'absence d'identifiant client est donc cohérente au niveau de la facture entière.

**Décision :**

- aucun identifiant client ne sera imputé artificiellement 
- les factures sans `Customer ID` seront conservées pour les analyses commerciales lorsqu'elles correspondent à des transactions valides 
- elles seront exclues des analyses nécessitant une identification individuelle du client 


In [56]:
# Nombre de lignes sans Customer ID par pays

display(
    missing_customer["Country"]
    .value_counts()
)

Country
United Kingdom          232320
EIRE                      1660
Hong Kong                  364
Unspecified                232
France                     128
Switzerland                125
Portugal                   116
United Arab Emirates       114
Bahrain                     67
Israel                      47
RSA                         46
Bermuda                     34
Sweden                      19
Lebanon                     13
Nigeria                      2
Name: count, dtype: int64

In [57]:
# Taux de Customer ID manquants par pays

missing_by_country = (
    missing_customer["Country"]
    .value_counts()
    .rename("missing_customer_rows")
    .to_frame()
)

all_by_country = (
    df_consolidated["Country"]
    .value_counts()
    .rename("total_rows")
    .to_frame()
)

missing_by_country = missing_by_country.join(all_by_country)

missing_by_country["missing_rate_pct"] = (
    missing_by_country["missing_customer_rows"]
    / missing_by_country["total_rows"]
    * 100
)

display(
    missing_by_country
    .sort_values("missing_customer_rows", ascending=False)
)

,missing_customer_rows,total_rows,missing_rate_pct
Country,,,
United Kingdom,232320,959983,24.200429
EIRE,1660,17689,9.384363
Hong Kong,364,364,100.000000
Unspecified,232,756,30.687831
France,128,14059,0.910449
Switzerland,125,3183,3.927113
Portugal,116,2540,4.566929
United Arab Emirates,114,500,22.800000
Bahrain,67,126,53.174603


### Conclusion sur les identifiants clients manquants

Le dataset consolidé contient 235 287 lignes sans `Customer ID`, soit 22,52 % des observations.

L'absence d'identifiant est cohérente au niveau des factures : aucune facture ne contient simultanément des lignes avec et sans `Customer ID`.

Le phénomène est très fortement concentré au Royaume-Uni, avec 232 320 lignes concernées, soit 24,20 % des lignes britanniques.

Certains marchés présentent également des taux élevés mais sur des volumes beaucoup plus faibles. Par exemple, Hong Kong ne contient aucun `Customer ID` renseigné dans les 364 lignes observées.

**Décision :**

- aucun `Customer ID` ne sera imputé artificiellement ;
- les transactions sans identifiant client seront conservées pour les analyses commerciales agrégées lorsqu'elles sont commercialement valides ;
- elles seront exclues des analyses individuelles clients ;
- la segmentation RFM, les cohortes, la rétention et les futurs modèles clients utiliseront uniquement les transactions disposant d'un `Customer ID` ;
- les comparaisons clients entre pays devront tenir compte des différences de taux d'identifiants manquants.

## Descriptions manquantes

In [ ]:
# Nombre de descriptions manquantes : 

missing_description = df_consolidated[
    df_consolidated["Description"].isna()
].copy()

print(
    "Nombre de lignes sans Description :",
    len(missing_description)
)

print(
    "Part du dataset :",
    round(len(missing_description) / len(df_consolidated) * 100, 2),
    "%"
)

Nombre de lignes sans Description : 4275
Part du dataset : 0.41 %


In [ ]:
# Les caractéristiques : 
print(
    "Quantity < 0 :",
    (missing_description["Quantity"] < 0).sum()
)

print(
    "Quantity > 0 :",
    (missing_description["Quantity"] > 0).sum()
)

print(
    "Price = 0 :",
    (missing_description["Price"] == 0).sum()
)

print(
    "Price > 0 :",
    (missing_description["Price"] > 0).sum()
)

print(
    "Customer ID manquant :",
    missing_description["Customer ID"].isna().sum()
)

Quantity < 0 : 2633
Quantity > 0 : 1642
Price = 0 : 4275
Price > 0 : 0
Customer ID manquant : 4275


In [60]:
display(
    missing_description[
        [
            "Invoice",
            "StockCode",
            "Quantity",
            "Price",
            "Customer ID",
            "Country",
            "InvoiceDate"
        ]
    ].head(30)
)

,Invoice,StockCode,Quantity,Price,Customer ID,Country,InvoiceDate
470,489521,21646,-50,0.0,NaN,United Kingdom,2009-12-01 11:44:00
3114,489655,20683,-44,0.0,NaN,United Kingdom,2009-12-01 17:26:00
3161,489659,21350,230,0.0,NaN,United Kingdom,2009-12-01 17:39:00
3731,489781,84292,17,0.0,NaN,United Kingdom,2009-12-02 11:45:00
4296,489806,18010,-770,0.0,NaN,United Kingdom,2009-12-02 12:42:00
4566,489821,85049G,-240,0.0,NaN,United Kingdom,2009-12-02 13:25:00
6378,489882,35751C,12,0.0,NaN,United Kingdom,2009-12-02 16:22:00
6555,489898,79323G,954,0.0,NaN,United Kingdom,2009-12-03 09:40:00
6576,489901,21098,-200,0.0,NaN,United Kingdom,2009-12-03 09:47:00
6581,489903,21166,48,0.0,NaN,United Kingdom,2009-12-03 09:57:00


In [61]:
display(
    missing_description["StockCode"]
    .astype(str)
    .value_counts()
    .head(30)
)

StockCode
84990     12
79321     11
22950     10
35965     10
23084     10
22139      9
22087      9
22084      9
71477      8
35970      7
21768      7
22528      7
POST       7
84795D     7
22451      7
37446      6
21478      6
21169      6
72803B     6
21340      6
84845C     6
20747      6
90084      6
84977      6
84467      6
37461      6
22467      6
20734      6
22689      6
21784      6
Name: count, dtype: int64

### Conclusion sur les descriptions manquantes

Le dataset consolidé contient 4 275 lignes sans `Description`, soit 0,41 % des observations.

L'analyse montre que :

- 2 633 de ces lignes présentent une quantité négative ;
- 1 642 présentent une quantité positive ;
- les 4 275 lignes ont un prix égal à zéro ;
- aucune ne possède un prix positif ;
- les 4 275 lignes ont également un `Customer ID` manquant.

Ces observations ne correspondent donc pas à des ventes directement valorisables dans les données disponibles.

**Décision :**

- les lignes seront conservées dans les données transformées afin de préserver la traçabilité ;
- elles seront identifiées comme lignes sans description ;
- elles seront exclues des indicateurs de chiffre d'affaires et de ventes produits ;
- elles ne seront pas utilisées dans les analyses clients ;
- aucune description ne sera imputée artificiellement.

## Analyse des doublons après consolidation

In [62]:
exact_duplicates = df_consolidated.duplicated().sum()

print(
    "Nombre de doublons exacts après consolidation :",
    exact_duplicates
)

Nombre de doublons exacts après consolidation : 11812


In [63]:
exact_duplicate_rows = df_consolidated[
    df_consolidated.duplicated(keep=False)
]

print(
    "Nombre total de lignes appartenant à un groupe dupliqué :",
    len(exact_duplicate_rows)
)

Nombre total de lignes appartenant à un groupe dupliqué : 22813


In [64]:
display(
    exact_duplicate_rows[
        [
            "Invoice",
            "StockCode",
            "Description",
            "Quantity",
            "InvoiceDate",
            "Price",
            "Customer ID",
            "Country"
        ]
    ]
    .sort_values(
        ["Invoice", "StockCode", "InvoiceDate"]
    )
    .head(50)
)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
379,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom
391,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom
365,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
386,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
363,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
371,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
394,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
362,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
385,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
368,489517,22130,PARTY CONE CHRISTMAS DECORATION,6,2009-12-01 11:34:00,0.85,16329.0,United Kingdom


In [65]:
print(
    "Factures concernées par les doublons exacts :",
    exact_duplicate_rows["Invoice"].nunique()
)

Factures concernées par les doublons exacts : 4387


In [66]:
print(
    "Part des doublons exacts :",
    round(
        exact_duplicates / len(df_consolidated) * 100,
        2
    ),
    "%"
)

Part des doublons exacts : 1.13 %


In [67]:
# combien de fois ? 
duplicate_group_sizes = (
    df_consolidated
    .groupby(
        [
            "Invoice",
            "StockCode",
            "Description",
            "Quantity",
            "InvoiceDate",
            "Price",
            "Customer ID",
            "Country"
        ],
        dropna=False
    )
    .size()
)

duplicate_group_sizes = duplicate_group_sizes[
    duplicate_group_sizes > 1
]

display(
    duplicate_group_sizes
    .value_counts()
    .sort_index()
)

2     10352
3       546
4        79
5        11
6        10
8         1
12        1
20        1
Name: count, dtype: int64

In [68]:
# impact : 
df_without_exact_duplicates = (
    df_consolidated
    .drop_duplicates()
)

print("Avant :", len(df_consolidated))
print("Après déduplication :", len(df_without_exact_duplicates))
print(
    "Lignes qui seraient retirées :",
    len(df_consolidated) - len(df_without_exact_duplicates)
)

Avant : 1044848
Après déduplication : 1033036
Lignes qui seraient retirées : 11812


In [69]:
print(
    "Réduction du volume :",
    round(
        (
            len(df_consolidated)
            - len(df_without_exact_duplicates)
        )
        / len(df_consolidated)
        * 100,
        2
    ),
    "%"
)

Réduction du volume : 1.13 %


### Conclusion sur les doublons

Après correction du chevauchement entre les deux feuilles, le dataset consolidé contient encore 11 812 doublons exacts, soit 1,13 % des observations.

Ces doublons concernent 4 387 factures.

La majorité des groupes dupliqués contient deux occurrences identiques, même si quelques groupes comportent davantage de répétitions.

Le dataset ne possède pas d'identifiant unique de ligne de facture. Il n'est donc pas possible de démontrer formellement que chaque répétition correspond à une erreur de saisie.

Cependant, les lignes concernées sont strictement identiques sur l'ensemble des attributs disponibles : facture, produit, description, quantité, date, prix, client et pays.

**Décision :**

- la couche Bronze conservera les données consolidées afin de préserver la traçabilité de la source ;
- la couche Silver supprimera les répétitions strictement identiques en conservant une seule occurrence ;
- cette règle retire 11 812 lignes, soit 1,13 % du dataset consolidé ;
- cette hypothèse sera documentée comme une règle de qualité des données.

## Synthèse de l'audit qualité

L'audit initial du dataset a permis d'identifier les principaux problèmes de qualité avant la construction du pipeline de données.

### 1. Chevauchement des sources

Les deux feuilles Excel contiennent une période commune du 1er au 9 décembre 2010.

22 523 lignes sont strictement identiques dans les deux feuilles.

**Règle :** une seule occurrence de cette période sera conservée.

### 2. Quantités négatives

22 557 lignes présentent une quantité négative après consolidation.

84,96 % appartiennent à une facture commençant par `C`, correspondant aux annulations.

Les autres lignes correspondent principalement à des mouvements tels que casse, perte, correction ou ajustement.

**Règle :** conserver les mouvements mais les classifier selon leur nature.

### 3. Prix négatifs

5 lignes présentent un prix négatif et correspondent à `Adjust bad debt`.

**Règle :** conserver mais identifier comme écritures non commerciales.

### 4. Prix nuls

6 024 lignes présentent un prix nul.

Une grande partie correspond à des corrections, mouvements internes ou observations non valorisées.

**Règle :** conserver mais ne pas considérer automatiquement comme ventes commerciales.

### 5. Prix extrêmes

La distribution des prix est fortement asymétrique et plusieurs prix élevés correspondent à des frais, commissions ou écritures manuelles.

**Règle :** aucun seuil arbitraire de prix ne sera utilisé pour supprimer les observations.

### 6. Customer ID manquant

235 287 lignes, soit 22,52 % du dataset, ne possèdent pas de `Customer ID`.

Aucune facture ne mélange lignes identifiées et non identifiées.

**Règle :**
- conserver ces transactions pour certaines analyses commerciales ;
- les exclure des analyses individuelles clients ;
- ne jamais imputer artificiellement un identifiant.

### 7. Description manquante

4 275 lignes, soit 0,41 %, ne possèdent pas de description.

Toutes ont un prix nul et un Customer ID manquant.

**Règle :** conserver pour la traçabilité mais exclure des analyses produits commerciales.

### 8. Doublons exacts

11 812 répétitions exactes subsistent après consolidation, soit 1,13 % du dataset.

**Règle :**
- conserver dans Bronze ;
- dédupliquer dans Silver en conservant une occurrence.